src/main/java/net/minecraftforge/common/ForgeHooks.java

```diff
@@ -1303,9 +1304,10 @@ public static String getRegistryName(Class<? extends TileEntity> type)
 
     public static boolean loadAdvancements(Map<ResourceLocation, Advancement.Builder> map)
     {
+        CraftingHelper.init();
```

Comment:
This seems out of place, won't it already be initialized by the static block?

Reply:
Hmm, is this the case if the log4j logger's debug is enabled by default? (If debug is enabled, advancements will be loaded in bootstrap)

Reply:
It needs to be called directly before loading advancements.

See `CraftingHelper.loadRecipes`:
https://github.com/MinecraftForge/MinecraftForge/blob/c5402d50fad5dbb649015d9b6d9bead6b710e035/src/main/java/net/minecraftforge/common/crafting/CraftingHelper.java#L606-L628

The static `init()` call in `CraftingHelper` is a red herring here, it is not relied upon and could probably just be removed (see #4014).

Note that advancements can be loaded multiple times - single initialisation is not appropriate as duplicated factory definitions are not allowed.

Reply:
I'm just not totally convinced that it's a good idea to make so many things public that aren't really designed to be. Perhaps things should be moved around to keep internal things internal.

Reply:
Well I did try to do things differently, but that ended up not working.

Do you have an idea for how this should be organised?

Otherwise I think reorganisation should be left for later, it can probably be done better by allowing breaking changes.